### PR TITLE
Li & Kim - Slack CLI

### PR DESCRIPTION
**Pull request by kimvitug05**
# Assignment Submission: Slack CLI

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
---      | ---
How did you go about exploring the Slack API? Did you learn anything that would be useful for your next project involving an API? | We explored the Slack API documentation and experimented with requests in Postman & HTTParty. Yes, don't hardcode your token and push to Github (your token will be revoked!!!).
Give a short summary of the _request/response cycle_. Where does your program fit into that scheme? | We send a request to Slack to get a list of all the `Channel`s and `User`s, or send a request to post a message. Slack gives a response of `ok` or `error` depending on whether the request is successful. 
How does your program check for and handle errors when using the Slack API? |  Our program will raise a `SlackApiError` if the HTTP request was not successful.
How did the design and organization of your project change over time? | We created the `User` and `Channel` class first. After Wave 2, we created the `Recipient` class, which `User` and `Channel` inherit from.
Did you use any of the _inheritance idioms_ we've talked about in class? How? | We utilized template methods in the `Recipient` class, which we implemented in `User` and `Channel`. We used `StandardError` as an abstract class.
How does VCR aid in testing a program that uses an API? | It records the response from the HTTP request and saves that data locally for later use.

### CODE REVIEW

File: lib/channel.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  attr_reader :topic, :member_count
+
+  def initialize(slack_id, name, topic, member_count)
+    super(slack_id, name)
+    @topic = topic
+    @member_count = member_count
+  end
+
+  def self.list_all
+    response = self.get("conversations.list")
```
Comment (CheezItMan) [inline lib/channel.rb:17]: I like the user of the helper method here!

File: lib/recipient.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    end
+
+    return response
+  end
+
+  def save_message_history(message)
+    CSV.open('history.csv', 'a') do |file|
+      file << [@name, @slack_id, message, Time.now] if message != ""
+    end
+  end
```
Comment (CheezItMan) [inline lib/recipient.rb:77]: Interesting way to save the message history.

File: lib/workspace.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
+require_relative 'user'
+require_relative 'channel'
+require 'httparty'
+require 'table_print'
+require 'dotenv'
+Dotenv.load
+
+class Workspace
```
Comment (CheezItMan) [inline lib/workspace.rb:8]: Good uses of `select` and `find` in this class.

File: test/recipient_test.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    end
+  end
+
+  describe "list_all" do
+    it "raises an NotImplementedError" do
+      expect{Recipient.list_all}.must_raise NotImplementedError
+    end
+  end
+
+  describe "send_message" do
```
Comment (CheezItMan) [inline test/recipient_test.rb:36]: You should also have a test where sending the message to an invalid recipient.

File: test/recipient_test.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    it "can send a super long message" do
+      VCR.use_cassette("slack-posts") do
+        response = @recipient.send_message("Hi You are beautiful!"*90000)
+        expect( response["ok"] ).must_equal true
+      end
+    end
+  end
+
+  describe "bot_post_message" do
```
Comment (CheezItMan) [inline test/recipient_test.rb:62]: You should also test posting an invalid message, edge cases.

File: test/recipient_test.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    it "can send a message as a bot" do
+      VCR.use_cassette("bot_post_message_recipient") do
+        response = @recipient.bot_post_message("Ting-Yi",":sparkles:", "Hello there!")
+        expect(response["ok"]).must_equal true
+        expect(response["message"]["text"]).must_equal "Hello there!"
+      end
+    end
+  end
+
+  describe "save_message_history" do
```
Comment (CheezItMan) [inline test/recipient_test.rb:76]: You should also have tests with invalid file names and with and without headers.

**Review by CheezItMan — COMMENTED** [review]
# Slack CLI



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices best practices working with APIs. The `.env` is not checked into git, and no API token was directly used in the Ruby code without `ENV`. | ✔️
Practices error handling with APIs. For all pieces of code that make an API call, it handles API requests that come back with errors/error status codes appropriately. | ✔️
Implements inheritance and inheritance idioms. There is a `Recipient` class. `User` and `Channel` inherit from `Recipient`. In `Recipient`, there are appropriate methods defined that are used in both `User` and `Channel`. Some may be implemented. Some may be template methods. | ✔️, nice work here!
Practices clean code. `lib/slack.rb` only interacts with `Workspace` to show a separation of responsibilities. Complex logic is broken into smaller helper methods. | ✔️, lots of good helper methods
Practices instance methods vs. class methods appropriately. The methods to list all Channels or Users is likely a class method within those respective classes. | ✔️
Practices best practices for testing. The project has and uses VCR mocking when running tests, and can run offline. | ✔️
Practices writing tests. The `User`, `Channel`, and `Workspace` classes have unit tests. | ✔️
Practices writing tests. There are tests for sending messages (the location of these tests may differ, but is likely in `Recipient`) | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
As a user of the CLI program, I can list users and channels | ✔️
As a user of the CLI program, I can select users and channels | ✔️
As a user of the CLI program, I can show the details of a selected user or channel | ✔️?
As a user of the CLI program, when I input something inappropriately, the program runs without crashing | ✔️?

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 7+ in Code Review && 3+ in Functional Requirements |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Summary

This was really excellent work.   You hit all the learning goals with a great design.  I like the extension for message history.  Take a look at my comments and let me know what questions you have.  Well done.
